Replicator/c4Replicator.hh

```diff
@@ -55,6 +55,12 @@ struct C4Replicator : public RefCounted,
 
     virtual void start() {
         LOCK(_mutex);
+        if(_status.level == kC4Stopping) {
+            Warn("Rapid call to start() (stop() is not finished yet), scheduling a restart after stop() is done...");
+            _cancelStop = true;
```

Comment:
In iOS's replicator, I ignored the start as the replicator has not been stopped yet. On the public facing side, users don't get STOPPED notification yet so I think it sort of expects that the start will be ignored.

Reply:
That is in no way clear from our contract, but are you asserting that the correct behavior is to ignore start unless the replicator is stopped?

Reply:
I do not depend on Core to do that for me.  On the other hand, that is exactly how the Java bindings behave.  You will never see two calls to stop, without a call to start in between, and vice-versa.

Reply:
I think it is worth to define the behavior. In other condition beside stopping, if the replicator is still running, calling start again will be no-ops. IMO, it should behave the same for stopping state. 

How will this affect the suspend and resume? In iOS, when the app goes to background, the replicator will be suspended. When the apps goes foreground, the replicator will be resume. There is also a case of switching b/w background and foreground is fast, the expectation is that the replicator should be resume when the app is foregrounded. When I implemented this in iOS, I did add an ability to cancel the suspend (auto resume).

Reply:
I took this implementation partially from the iOS implementation and spent quite a long time thinking about these situations.  If a rapid foreground is expected to resume the replicator from suspending then why is a rapid start not expected to behave similarly?  Here are the assumptions I am working on:

- Calling start when:
    - stopped -> start the replicator immediately
    - stopping -> start the replicator after it is stopped (should be ignore?)
    - all others -> ignore
- Calling stop when:
    - stopped -> ignore
    - stopping -> cancel any resume efforts, if applicable
    - offline -> switch to stopped immediately
    - all others -> transition to stopping and start the stop process
- Backgrounding when:
    - stopped or stopping (or connecting ?) -> ignore
    - All others -> suspend
 - Foregrounding when:
    - stopping -> if suspended flag is set, set resume flag otherwise ignore
    - offline -> resume
    - all others -> ignore

It's pretty frustrating that this has only been discussed now after a good deal of implementation has already been done (previous to even this PR).  Having the start be a no op for the stopping state would make me happy because it is easier to implement, but it also presents an inconsistency with the way rapid backgrounding and foregrounding is handled (I guess this could be explained away since those calls are automatic versus the other calls which are explicit)